src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/ProjectSystem/RemoteProjectSnapshot.cs

```diff
@@ -43,26 +44,28 @@ public RemoteProjectSnapshot(Project project, DocumentSnapshotFactory documentSn
         _telemetryReporter = telemetryReporter;
         Key = _project.ToProjectKey();
 
-        _lazyConfiguration = new Lazy<RazorConfiguration>(CreateRazorConfiguration);
-        _lazyProjectEngine = new Lazy<RazorProjectEngine>(() =>
+        _lazyConfiguration = new AsyncLazy<RazorConfiguration>(CreateRazorConfigurationAsync, joinableTaskFactory: null);
+        _lazyProjectEngine = new AsyncLazy<RazorProjectEngine>(async () =>
         {
+            var configuration = await _lazyConfiguration.GetValueAsync();
             return ProjectEngineFactories.DefaultProvider.Create(
-                _lazyConfiguration.Value,
+                configuration,
                 rootDirectoryPath: Path.GetDirectoryName(FilePath).AssumeNotNull(),
                 configure: builder =>
                 {
                     builder.SetRootNamespace(RootNamespace);
                     builder.SetCSharpLanguageVersion(CSharpLanguageVersion);
                     builder.SetSupportLocalizedComponentNames();
                 });
-        });
+        },
+        joinableTaskFactory: null);
 
         _importsToRelatedDocumentsLazy = new Lazy<ImmutableDictionary<string, ImmutableArray<string>>>(() =>
         {
             var importsToRelatedDocuments = ImmutableDictionary.Create<string, ImmutableArray<string>>(FilePathNormalizingComparer.Instance);
             foreach (var documentFilePath in DocumentFilePaths)
             {
-                var importTargetPaths = ProjectState.GetImportDocumentTargetPaths(documentFilePath, FileKinds.GetFileKindFromFilePath(documentFilePath), _lazyProjectEngine.Value);
+                var importTargetPaths = ProjectState.GetImportDocumentTargetPaths(documentFilePath, FileKinds.GetFileKindFromFilePath(documentFilePath), _lazyProjectEngine.GetValue());
```

Comment:
Do you mean to synchronously block on the `AsyncLazy` here? What is the impact of that?

Reply:
I did mean to do that. I don't foresee it being an issue because the blocking would be something that's free threaded (waiting on a compilation)

Would a better option be to make this `AsyncLazy` as well?

Reply:
Your thoughts here are totally reasonable. I don't know the order of everything, but will `GetRelatedDocuments()` or `IsImportDocument()` (the methods that realizes this lazy) always be called after `_lazyProjectEngine` has been computed? If so, would a `Debug.Assert(_lazyProjectEngine.IsValueFactoryCompleted)` be helpful?

I searched for usages of `GetRelatedDocuments` and `IsImportDocument`, and I think that assert would hold true on our current code:

1. `GetRelatedDocuments` - Not used by `MS.CA.Remote.Razor` or any of its dependencies, so it is never called on a `RemoteProjectSnapshot`. This is only called in the language server and in Visual Studio.
2. `IsImportDocument` - This is called by a single extension method that _is_ used by `MS.CA.Remote.Razor`. However, it calls `IDocumentSnapshot.GetGeneratedOutputAsync()` before calling `IsImportDocument`. So, the project engine will have been retrieved before accessing the `_importsToRelatedDocumentsLazy`.

Just a thought. Totally up to you.